### ISSUE TITLE
CN: adjust behaviour `--skip`, `--only`, consider `--no-transitive-deps`

### ISSUE THREAD

**Issue by peterohanley** [rems-project/cerberus#314]
Here CN complains that there is no `Owned<...>` available for `g` because `a` is only `Block<...>`. In the example before I reduced it there was code in `f` that initialized `a`. If `f` is only a declaration there is the correct error.

Allowing functions whose body is in scope looks like it was an intentional feature but it is unsafe.

```
// cn --magic-comment-char-dollar no_check_for_unspec.c --only=h
static void
f(char a[1])
{
}

static void
g(char a[1])
/*$
  requires take tin = each(u64 i; i < 1u64) {Owned<char>(array_shift(a, i))};
  ensures take tout = each(u64 i; i < 1u64) {Owned<char>(array_shift(a, i))};
$*/
{
}

static void
h(void)
/*$
  requires true;
  ensures true;
 $*/
{
    char a[1];
    f(a);
    g(a);
}
```

**Comment by dc-mak** [comment]
I'm not exactly sure what you mean "unsafe" here? 

Even with `char a[1] = {'a'};`, I agree it looks a bit suspicious if `h` passes type but as far as CN is concerned, no spec means it's not doing anything observable. It is intentional, but I don't see how it is  unsafe.

If `f` was writing to `a[0]`, then it would fail type-checking when checked against no spec (due to lack of ownership). And if the spec were to change, then `g` or `h` would need adjusting accordingly.

Perhaps there's a UI question here on communicating to users that even if 999/1000 function verify, that gives no guarantee about running the code with the 1000th function?

**Comment by peterohanley** [comment]
>as far as CN is concerned, no spec means it's not doing anything observable. It is intentional, but I don't see how it is unsafe.

> Perhaps there's a UI question here on communicating to users that even if 999/1000 function verify, that gives no guarantee about running the code with the 1000th function?

These are incompatible. The latter requires no spec mean anything can happen. In this example, CN returns a warning that a resource is not available for `g`. No mention was made of `f` but we don't even know if the program will still exist after calling a mystery meat function. I tried with `exit` called from `f` and the same error is produced.

If CN is inferring specs for functions with available bodies (it seems it must, since it does error if `f` is only a declaration) I would like to see them and if it is not it should not silently allow calling them. A warning about a spec violation after calling one is misleading.

**Comment by thatplguy** [comment]
@peterohanley I don't quite follow.  Which function is the mystery function in this example?  Is the issue that it's unclear what CN does when no specification is attached to a function definition?

AFAIK omitting a specification is the same as supplying an empty specification – the function has no assumptions needed to avoid undefined behavior, nor does it allocate memory that it then relinquishes ownership of.  Maybe we can surface this kind of information in the IDE, e.g. mousing over a function will show the CN spec for it, or a brief description of the empty spec.

To make sure we're on the same page, here's what I would expect to happen for the example you listed (@dc-mak correct me if I get this wrong):
1. CN attempts to verify `h`, as per the `--only=h` flag.
2. CN verifies functions used in `h`, which is necessary to verify `h`.
    1. CN verifies that `f` satisfies the empty spec, which means that it doesn't read/write `a` or otherwise do anything resulting in undefined behavior (like divide by a var that could be zero).
    1. CN verifies that `g` satisfies its spec.
3. CN verifies that `h` satisfies the preconditions for `f` and `g` as part of verifying `h`.  This succeeds for `f` because it has no requirements, but it fails with an error because `g` requires `Owned<char>(&a + i)`.

**Comment by peterohanley** [comment]
I think by reducing this test case to an empty function I made this more confusing so here are two specific examples. Note I have removed the `--only` flag, which I added as part of minimizing the error, but because of the order CN decides to process the functions the error is the same.

# 1 nasal demons

I run CN on this file:

```
#include <stdlib.h>
static void
f(char a[1])
{
exit(1);
}

static void
g(char a[1])
/*$
  requires take tin = each(u64 i; i < 1u64) {Owned<char>(array_shift(a, i))};
  ensures take tout = each(u64 i; i < 1u64) {Owned<char>(array_shift(a, i))};
$*/
{
}

static void
h(void)
/*$
  requires true;
  ensures true;
 $*/
{
    char a[1];
    f(a);
    g(a);
}
```

```
guso@000492-guso src % cn --magic-comment-char-dollar no_check_for_unspec.c
[1/1]: h
no_check_for_unspec.c:27:5: error: Missing resource for calling function g
    g(a);
    ^~~~ 
Resource needed: each(u64 i; i < 1u64)
{Owned<char>(&a + i * 1u64)}
      no_check_for_unspec.c:12:17: (arg tin)
Consider the state in /var/folders/tz/g0kxkm2915xdtm6qgv8rff000000gp/T/state_a7b07f.html
```

I think this is bad. `f` doesn't even return. The resources available after the silently accepted call to `f` are not trustworthy. I think the error should be no spec for `f`, even when only processing `h`.

# 2 the initial case

```
// cn --magic-comment-char-dollar veiled_initalization.c
void e(char a[1]);
/*$ spec e(pointer a);
  requires take tin = each(u64 i; i < 1u64) {Block<char>(array_shift(a, i))};
  ensures take tout = each(u64 i; i < 1u64) {Owned<char>(array_shift(a, i))};
$*/

static void
f(char a[1])
{
    e(a);
}

static void
g(char a[1])
/*$
  requires take tin = each(u64 i; i < 1u64) {Owned<char>(array_shift(a, i))};
  ensures take tout = each(u64 i; i < 1u64) {Owned<char>(array_shift(a, i))};
$*/
{
}

static void
h(void)
/*$
  requires true;
  ensures true;
 $*/
{
    char a[1];
    f(a);
    g(a);
}
```
```
guso@000492-guso src % cn --magic-comment-char-dollar veiled_initalization.c          
[1/1]: h
veiled_initalization.c:36:5: error: Missing resource for calling function g
    g(a);
    ^~~~ 
Resource needed: each(u64 i; i < 1u64)
{Owned<char>(&a + i * 1u64)}
      veiled_initalization.c:21:17: (arg tin)
Consider the state in /var/folders/tz/g0kxkm2915xdtm6qgv8rff000000gp/T/state_6b039e.html
```

Here the issue is `f` is missing the spec that passes on the effects of `e`. This is how I noticed this.

If I pass `--skip=h` to CN for both files I get the following:
```
guso@000492-guso src % cn --magic-comment-char-dollar no_check_for_unspec.c --skip=h
[1/2]: g
[2/2]: f
no_check_for_unspec.c:6:1: error: Call to function with no spec: exit
exit(1);
^~~~~~~ 
```
```
guso@000492-guso src % cn --magic-comment-char-dollar veiled_initalization.c --skip=h
[1/2]: g
[2/2]: f
veiled_initalization.c:15:5: error: Missing resource for calling function e
    e(a);
    ^~~~ 
Resource needed: each(u64 i; i < 1u64)
{Block<char>(a + i * 1u64)}
      veiled_initalization.c:5:17: (arg tin)
Consider the state in /var/folders/tz/g0kxkm2915xdtm6qgv8rff000000gp/T/state_f6c72d.html
```

Both errors are about `f`, as I would like.

So I think the issue here is actually that 2 and 3 in your list happen in opposite orders.

**Comment by thatplguy** [comment]
Thanks @peterohanley, I think I understand.  Let me see if I can restate the issue:

There are at least two errors in each example, and CN is picking the wrong one to show.  In both cases, the error reported is the last error in the execution trace, when in fact it's only an error because of earlier errors.
* In the first example, execution can't actually reach `g`, and 
* in the second example, `f` shouldn't be able to run because it doesn't establish the necessary ownership.  

A better solution would report these errors first, and only report an error for `g` if execution can successfully reach it.  Does that sound like the right high-level summary?

Related to your earlier example (attn @dc-mak) the `--only` and `--skip` options seem dangerous.  From what I can tell, excluded functions are *assumed* to match their specs, or the empty spec if no spec is present.  Is that right @dc-mak?

This gives the impression that `h` verifies in this example, when in fact it only verifies if one (erroneously) assumes that all its dependencies are correct:

```
// cn ex2.c --only=h

static void inc(char a[], unsigned int len)
{
    for (unsigned int i = 0; i < len; i++)
    {
        a[i] += 1;
    }
}

static void h(void)
{
    char a[1];
    inc(a, 1);
}
```

A better solution would either be to fail if a dependent function is skipped (via `--only` or `--skip`) or perhaps change the output to indicate conditional success, e.g. something like

```
> cn ex2.c --only=h
[1/1]: h (CONDITIONAL)
[1/1]: h - WARNING: h depends on inc, which was not verified.  Remove the `--only` flag for full verification.
```

**Comment by PeterSewell** [comment]
- CN is doing function-local verification: when verifying a function, it
uses the specs of other functions, not their implementations.  Here the
spec of f is just the implicit spec that it doesn't have UB

- it's a logic of partial correctness, not total correctness - I don't
think there's a way of telling it that some function doesn't return
(though @Christopher
Pulte ***@***.***> @Dhruv Makwana ***@***.***> @Neel Krishnaswami
***@***.***> should comment)

**Comment by dc-mak** [comment]
> In the first example, execution can't actually reach g, and

This what we get when we have a partial-correctness logic as CN is.

>    in the second example, f shouldn't be able to run because it doesn't establish the necessary ownership.

There is no necessary ownership to be established. `char a[1]` is just a char pointer which is unused, passed as a value. You would need ownership to dereference or write to `a` but as the program stands, there is no ownership problem. The issue with program two is actually in `g` as the error reports:

```
Requested:
each(u64 i; i < 1u64) {Owned<char>(&a + i * 1u64)}

Available resources:
each(u64 i; (0u64 <= i) && (i < 1u64)) {Block<char>(&a + i * 1u64)}(default(map<u64, u8>))
```

The solution is to initialise `a` with `char a[1] = {'a'};`.

> Related to your earlier example (attn @dc-mak) the `--only` and `--skip` options seem dangerous. From what I can tell, excluded functions are _assumed_ to match their specs, or the empty spec if no spec is present. Is that right @dc-mak?

This is correct.

> This gives the impression that `h` verifies in this example, when in fact it only verifies if one (erroneously) assumes that all its dependencies are correct:

Correct.

> A better solution would either be to fail if a dependent function is skipped (via `--only` or `--skip`) or perhaps change the output to indicate conditional success, e.g. something like

> ```
> > cn ex2.c --only=h
> [1/1]: h (CONDITIONAL)
> [1/1]: h - WARNING: h depends on inc, which was not verified.  Remove the `--only` flag for full verification.
> ```

Changing the output is reasonable. I understand that partial correctness and per-function checking are confusing in this instance, but changes to both would require revisiting CN design choices fundamentally. Per-function checking is *essential* for scaling to large code bases so that checking can be parallelised.

Another feature would be to have `--transitive-deps-of` flag to check a function and its dependencies. And in general, we have been wanting to have CN check functions in topological order for a while (leaves first).

I also think it's reasonable to revisit the choice of warning when a declared function has no spec vs defined function which is just assumed to be no spec.

I appreciate that CN is very rough unpolished research software with many quirks and unexplained design choices, leading to much frustration and so concrete suggestions for improvements are well received, _provided users take the time to ask questions and understand the current state of affairs first_.

**Comment by thatplguy** [comment]
Thanks @dc-mak.  I appreciate the time you're taking to help us understand these issues.  I should have written "I had expected to see" rather than "A better solution".  My intent was to make my confusion more clear, not advocate for a particular path forward.

> Per-function checking is essential for scaling to large code bases so that checking can be parallelised.

I don't think we would need to move away from per-function checking.  My concern is really just with the `--only` and `--skip` flags.  They seem useful as a CN developer, but subtle as a user.  Here are some scenarios where I think it would be easy for a user to make a mistake:
1. I'm adding specs gradually to my codebase, starting with `f1`.  I run CN with `--only=f1`, and it verifies, so I think that my code is correct, but in fact it calls other side-effecting functions that don't have specs.
2. I'm trying to verify a complicated function and it's not working, so I use `--skip` to temporarily ignore it for now.  However, all the functions that call this function now appear to be verified.

Something like `--transitive-deps-of` seems like an excellent solution that doesn't require changing anything fundamental about CN, although perhaps it should be turned on by default when using `--skip` and `--only`, and one could use `--no-transitive-deps-of` to recover the current behavior?

Not a high-priority change to make, but something to consider before the end of P1.

> The solution is to initialise a with char a[1] = {'a'};.

In this example ("2 the initial case"), the function `e` is intended to initialize `a`, but the spec for `f` doesn't capture that effect.  Although `char a[1] = {'a'};` would work, I think the fix @peterohanley actually needed to make was to add a spec to `f` that requires `block` for its argument and ensures `owned`.

In this case, there are two errors that could be reported – the one in `g`, but also the call to `e` in `f`, which should fail because calling `e` requires `block(a)`.  Fixing the second error (assuming you also add an `ensures` clause when adding the `requires` clause) would fix the first error.

This kind of problem has turned up before in the type checking literature – e.g. [this paper](https://www.cs.cornell.edu/andru/papers/diagnostic/popl14.pdf).  I don't think this is a top priority right now, but in the future it might be interesting to explore heuristics for how CN reports errors.

## Popping up a level

@peterohanley, thanks for reporting this – it's very helpful to see these points of confusion, and indeed it uncovered a few points I was confused about as well.

To summarize, here's what I see as relevant to this issue:
1. There is no CN bug to fix in a technical sense.  This is a usability issue.
2. Reporting more than one error might help (related to #287).
3. We should add something to the tutorial about partial correctness – the fact that CN doesn't reason about termination is normal for these tools but surprising if you don't come from a verification background.  In the future, maybe there are heuristics we can adopt that help, e.g. adding library models for `exit` and the like that factor into the analysis.
4. We should also add something about how functions without specs are treated.
5. @peterohanley you should probably stop using `--only` and `--skip`, unless you're confident they're doing what you want 🙂

Thoughts?

**Comment by peterohanley** [comment]
Changing the output or checking in bottom-up order would (each) resolve this.

>  The issue with program two is actually in g as the error reports: ... The solution is to initialise a with char a[1] = {'a'};.

This is in a large file that is still having specs added, and the solution was to add a spec for `f`. I opened this issue because a warning on the call to `f` in this situation would help point me in the right direction.

**Comment by peterohanley** [comment]
> 1. There is no CN bug to fix in a technical sense.  This is a usability issue.

Agreed.


> 3. We should add something to the tutorial about partial correctness – the fact that CN doesn't reason about termination is normal for these tools but surprising if you don't come from a verification background.  In the future, maybe there are heuristics we can adopt that help, e.g. adding library models for `exit` and the like that factor into the analysis.

This was a digression that I thought would help explain why I wanted the warning in general but it wasn't helpful.

CN actually handles noreturn pretty well:
```
void exit(int x);
/*$ spec exit(i32 x);
    requires true;
    ensures false;
$*/

static void
h(void)
/*$
  requires true;
  ensures true;
 $*/
{
    exit(1);
}
```
```
guso@000492-guso src % cn --magic-comment-char-dollar exit.c 
[1/1]: h
veiled_initalization.c:15:5: warning: Completed type-checking early along this path due to inconsistent facts.
    exit(1);
    ^~~~~~~~ 
```

> 2. Reporting more than one error might help (related to [[CN] `cn` can't report rich errors in multiple unrelated functions #287](https://github.com/rems-project/cerberus/issues/287)).

> 5. @peterohanley you should probably stop using `--only` and `--skip`, unless you're confident they're doing what you want 🙂

 I use `--skip` as a workaround for 2. above. I used `--only` to ensure the specific error message was produced in my test case and it caused a lot of confusion. I don't use it normally.

**Comment by peterohanley** [comment]
A skipped function with an inconsistent assumptions error still throws an error. `--skip` seems to be checked too late in the processing to stop this.
